### ISSUE TITLE
Add Line Breaks

### ISSUE THREAD

**Issue by tobi-coder** [EvotecIT/PSWriteOffice#5]
Hi. What is the correct way to add a line break / new line? I didn't find anything in the examples.

**Comment by PrzemyslawKlys** [comment]
$Document.AddLineBreak() - as far as I remember - I guess someone should help me out with converting this to PowerShell commands ;)

**Comment by tobi-coder** [comment]
I tried that without success.

```
$filePath = "$Env:USERPROFILE\Desktop\$([Guid]::NewGuid()).docx"
$document = [OfficeIMO.Word.WordDocument]::Create($filePath)
$document.AddParagraph("Row1")
$document.AddParagraph("Row2")
$Document.AddLineBreak()
$Document.AddLineBreak()
$Document.AddLineBreak()
$document.AddParagraph("Row3")
$document.Save()
```

It returns:
```
Row1
Row2
Row3
```

but it should be

```
Row1
Row2


Row3
```

**Comment by PrzemyslawKlys** [comment]
Sorry.

```powershell
$Document.AddParagraph()
$Document.AddParagraph()
$Document.AddParagraph()
$Document.AddParagraph()
```

Will work adding an empty line.

```powershell
$Document = New-OfficeWord -FilePath $PSScriptRoot\Documents\DocList.docx

New-OfficeWordText -Document $Document -Text 'This is a test, very big test ', 'and this should be bol' -Bold $null, $true -Underline $null, 'Dashed'

$List = New-OfficeWordList -Document $Document {
    New-OfficeWordListItem -Text 'Test1'
    New-OfficeWordListItem -Text 'Test2'
    New-OfficeWordListItem -Text 'Test3' -Level 2
} -Style Heading1ai

$Document.AddParagraph()
$Document.AddParagraph()
$Document.AddParagraph()
$Document.AddParagraph()


$P1 = New-OfficeWordListItem -Text 'Test4' -List $List

New-OfficeWordText -Document $Document -Text "But lists don't really have to be next to each other" -Bold $true -Alignment Center -Color RoseBud

$P2 = New-OfficeWordListItem -Text 'Test5' -List $List

New-OfficeWordText -Document $Document -Text "Here's another way to define list" -Bold $true -Alignment Center -Color RoseBud

$List1 = $Document.AddList([OfficeIMO.Word.WordListStyle]::Headings111)
$Paragraph1 = $List1.AddItem("Test", 2)
$Paragraph1.FontSize = 24
$Paragraph2 = $List1.AddItem("Test1", 1)
$Paragraph2.Strike = $true
$Paragraph3 = $List1.AddItem("Test2", 0)
$Paragraph3 = $Paragraph3.SetColor([SixLabors.ImageSharp.Color]::Red)

$null = $Document.AddPageBreak()

$P3 = New-OfficeWordListItem -Text 'Test6' -List $List
$P3.Bold = $true
$P3.FontSize = 20

# Get all listitems and find the listitem you want to change
$List.ListItems | Format-Table
$List.ListItems[0].Bold = $true

Save-OfficeWord -Document $Document -Show -FilePath $PSScriptRoot\Documents\Doc1Updated.docx
```

I've confused it with PageBreak.

**Comment by GhostyIs1337** [comment]
How about adding line breaks to text inside tables? :)

**Comment by PrzemyslawKlys** [comment]
It's a few versions behind, but this could work

```powershell
$table.Rows[($y + $x)].Cells[0].Paragraphs[0].AddParagraph()
```

**Comment by GhostyIs1337** [comment]
```ps1
Method invocation failed because [OfficeIMO.Word.WordParagraph] does not contain a method named 'AddParagraph'. 
```

:)

I've figured it out though, I was adding .AddParagraphAfterSelf() and .AddParagraphBeforeSelf() in reverse.

This works fine now though:
```ps1
            $table.Rows[($y + $x)].Cells[0].MergeHorizontally(6, 0)
            $table.Rows[($y + $x)].Cells[0].Paragraphs[0].AddParagraphAfterSelf()
            $table.Rows[($y + $x)].Cells[0].Paragraphs[0].Text = "$($integration.Purpose)"
            $table.Rows[($y + $x)].Cells[0].Paragraphs[0].ParagraphAlignment = "Center"
            $table.Rows[($y + $x)].Cells[0].Paragraphs[0].AddParagraphBeforeSelf()
```

Now just have to figure out how to add it to $OFS :D

**Comment by GhostyIs1337** [comment]
@PrzemyslawKlys 
Hey, do you have some input? I can't seem to figure out how to add line breaks here:
```ps1
$inboundStringBuilder = [System.Text.StringBuilder]""
$commentStringBuilder = [System.Text.StringBuilder]""

foreach ($jsonBuild in ($integration.Rule | where { $_.Type -eq "Inbound" })) {
    if ($jsonBuild.Switchable -eq "True") {
        $null = $commentStringBuilder.AppendLine("$($jsonBuild.Type)`r`nSWITCHABLE`r`n`r`n$($jsonBuild.Comment)")
    }
    else {
        $null = $commentStringBuilder.AppendLine("$($jsonBuild.Type)`r`n`r`n$($jsonBuild.Comment)")
    }
    foreach ($inRule in $jsonBuild.Destination) {
         $null = $inboundStringBuilder.AppendLine("$($inRule.Protocol)://$($inRule.Domain)$($inRule.Path)`r`nPoint to $($inRule.Host)`r`n`r`n" )
     }
}
                

$inboundEndpoint = $inboundStringBuilder.ToString()
$inboundComment = $commentStringBuilder.ToString()
```

And to $OFS:

```ps1
$OFS = "`r`n"
# Other code that uses it, foreach pipelines etc
# -ComObject example:
$Table.cell(($a + 2), 1).range.text = $integration.Rule | where { $_.Type -eq "Inbound" } | foreach { $_.SourceIP } | Sort-Object | Get-Unique | Out-String

#OfficeIMO:
$table.Rows[($a + 2)].Cells[0].Paragraphs[0].Text = $integration.Rule | where { $_.Type -eq "Inbound" } | foreach { $_.SourceIP } | Sort-Object | Get-Unique | Out-String

Remove-Variable OFS
```

I tried adding .AddParagraphBefore/AfterSelf(), but it just adds one before/after the blob of text that gets put into the cell, not between all of it.

This is text that goes into cells all over the table, but since it's multiple objects being passed from the JSON to one cell, it all gets written to one line inside the cell. The above code works with -ComObject Word.Application - I know it doesn't matter much, just as an input.

**Comment by PrzemyslawKlys** [comment]
Adding "`r`n" or anything like that won't work because it's not doing anything in the Text.

![image](https://user-images.githubusercontent.com/15063294/199715793-0bc700e1-05c6-4149-bbfd-e55403399dc3.png)

I've now tested 2 solutions. One is something I've added just now, the other one is something that's builtin and it seems to work. 

![image](https://user-images.githubusercontent.com/15063294/199715909-fb37869d-2f88-4fed-b79c-d4ebf9f9b0a8.png)

In other words you were not working with paragraphs you have added, but with first paragraph of each cell. When you do AddParagraphAfterSelf or AddParagraphBefore() you need to make sure you are updating variable with new paragraph as new paragraph is returned.

![image](https://user-images.githubusercontent.com/15063294/199716306-2bb1e125-f22b-4be6-9f0d-2bbb1ae9a02a.png)

**Comment by PrzemyslawKlys** [comment]
You can also do this:

![image](https://user-images.githubusercontent.com/15063294/199717087-e47af660-4ad7-43dc-9dfc-d815b4b31efd.png)

**Comment by PrzemyslawKlys** [comment]
Also keep in mind the AddBreak and AddParagraph will have a different behavior. If what you want is AddBreak then we need to "update the DLL" in this project and release new version, but the idea is the same.

**Comment by GhostyIs1337** [comment]
Interesting approach, I'll have to play around with it.

However, I think AddBreak() would be more inline with what I'm trying to achieve, especially since I have to remove some empty rows and looking through empty paragraphs is sort of annoying to determine throughout the whole table which are actually empty and which aren't. :)

Keeping the .dll up to date with the main one, especially on major changes probably wouldn't be a bad idea either - but that is entirely up to you, of course.

**Comment by PrzemyslawKlys** [comment]
I guess it would be possible to write that logic that if you use "\r\n" in text it should add new paragraph automatically.

**Comment by GhostyIs1337** [comment]
Something to look at for sure. :)

**Comment by v-bulynkin** [comment]
How to add not a paragraph but a new line in the same paragraph? It is indicated by the Enter symbol below and can be typed by Shift+Enter keys.

![изображение](https://user-images.githubusercontent.com/56349712/222903408-35cdb60f-26f4-46e1-8e9f-f18594c43a6d.png)

**Comment by PrzemyslawKlys** [comment]
```powershell
$paragraph2 = $document.AddParagraph("Adding paragraph2 with some text and pressing SHIFT+ENTER");
paragraph2.AddBreak();
paragraph2.AddText("Continue1");
paragraph2.AddBreak();
paragraph2.AddText("Continue2");
```

**Comment by v-bulynkin** [comment]
Thank you! However, I don't clearly understand how to combine it with the syntax
```
New-OfficeWordText -Document $doc -Alignment Center -Text "Adding paragraph2 with some text and pressing SHIFT+ENTER"
```
What if I need to make one line bold or italic if I use ```paragraph2.AddText("Continue1");```?

**Comment by PrzemyslawKlys** [comment]
```powershell
$Document = New-OfficeWord -FilePath $PSScriptRoot\Documents\BasicDocument.docx

New-OfficeWordText -Document $Document -Text 'This is a test, very big test ', 'and this should be bold' -Bold $null, $true -Underline Dash, $null

New-OfficeWordText -Document $Document -Text 'This is a test, very big test', 'ooops' -Color Blue, Gold -Alignment Right

$Paragraph = New-OfficeWordText -Document $Document -Text 'Centered' -Color Blue, Gold -Alignment Center -ReturnObject
$Paragraph = $Paragraph.AddBreak()
New-OfficeWordText -Document $Document -Text ' Attached to existing paragraph', ' continue' -Paragraph $Paragraph -Color Blue

Save-OfficeWord -Document $Document -Show
```

This should work. So you have to use ReturnObject and you can mix and match ways to build things. You can also just work with the native way

```powershell
$Document = New-OfficeWord -FilePath $PSScriptRoot\Documents\BasicDocument.docx

New-OfficeWordText -Document $Document -Text 'This is a test, very big test ', 'and this should be bold' -Bold $null, $true -Underline Dash, $null

New-OfficeWordText -Document $Document -Text 'This is a test, very big test', 'ooops' -Color Blue, Gold -Alignment Right

$Paragraph = New-OfficeWordText -Document $Document -Text 'Centered' -Color Blue, Gold -Alignment Center -ReturnObject
$Paragraph = $Paragraph.AddBreak()
$Paragraph = New-OfficeWordText -Document $Document -Text ' Attached to existing paragraph', ' continue' -Paragraph $Paragraph -Color Blue -ReturnObject
$Paragraph.Bold = $true
$Paragraph = $Paragraph.SetItalic().AddText("More txt").SetBold()
$Paragraph = $Paragraph.AddText("Even more text").AddBreak()
$Paragraph.AddText("Mix and match").SetBold().SetItalic()
$Paragraph = $Paragraph.SetUnderline([DocumentFormat.OpenXml.Wordprocessing.UnderlineValues]::Dash)
Save-OfficeWord -Document $Document -Show
```

Normally I guess we should add `New-OfficeWordBreak` and have a nice little helper function, but building both OfficeIMO and PSWriteHTML and all other modules doesn't give me much time.

**Comment by v-bulynkin** [comment]
Works great, thank you very much for the explanation!

**Comment by PrzemyslawKlys** [comment]
I prepared PR #85 for this. It adds `Add-OfficeWordBreak` plus the `WordBreak` alias for same-paragraph line breaks, while still allowing native OfficeIMO chaining through `-PassThru`.

```powershell
New-OfficeWord -Path .\breaks.docx {
    Add-OfficeWordParagraph {
        Add-OfficeWordText 'Line 1'
        Add-OfficeWordBreak
        Add-OfficeWordText 'Line 2'
    }
}
```

The PR uses `Fixes #5`, so this will close when merged.
